src/osd/DynamicPerfStats.h

```diff
@@ -204,30 +205,38 @@ class DynamicPerfStats {
           continue;
         }
 
+        // Weighted Random Sampling (Algorithm A-Chao):
+        // Select the first [0, max_count) samples, randomly replace
+        // with samples from [max_count, end) using weighted
+        // probability, and return [0, max_count) as the result.
+
+        ceph_assert(limit.max_count < counters.size());
         typedef std::map<OSDPerfMetricKey, PerformanceCounters>::iterator
             Iterator;
         std::vector<Iterator> counter_iterators;
-        counter_iterators.reserve(counters.size());
-        for (Iterator it_counters = counters.begin();
-             it_counters != counters.end(); it_counters++) {
-          counter_iterators.push_back(it_counters);
+        counter_iterators.reserve(limit.max_count);
+
+        Iterator it_counters = counters.begin();
+        uint64_t wsum = 0;
+        for (size_t i = 0; i < limit.max_count; i++) {
+          wsum += it_counters->second[index].first;
+          counter_iterators.push_back(it_counters++);
+        }
+        for (; it_counters != counters.end(); it_counters++) {
+          wsum += it_counters->second[index].first;
+          if (ceph::util::generate_random_number(0, wsum) <=
```

Comment:
You should use a class-level instance of ```random_number_generator``` and use its functor operator to generate each number. Using the method directly will just initialize a new rng engine for each function call.

Reply:
@dillaman I thought `ceph::util::generate_random_number` will reuse `thread_local rng_engine` form  "include/random.h"? No?

Reply:
Missed that implementation detail -- thx.